charts/replicated-library/templates/lib/_statefulset.tpl

```diff
@@ -0,0 +1,68 @@
+{{/*
+This template serves as the blueprint for the StatefulSet objects that are created
+within the common library.
+*/}}
+{{- define "replicated-library.statefulset" }}
+  {{- $values := . -}}
+  {{- if hasKey . "AppValues" -}}
+    {{- with .AppValues.app -}}
+      {{- $values = . -}}
+    {{- end -}}
+  {{ end -}}
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ include "replicated-library.names.appname" . }}
+  {{- with (merge ($values.labels | default dict) (include "replicated-library.labels" $ | fromYaml)) }}
+  labels: {{- toYaml . | nindent 4 }}
+  {{- end }}
+  {{- with (merge ($values.annotations | default dict) (include "replicated-library.annotations" $ | fromYaml)) }}
+  annotations: {{- toYaml . | nindent 4 }}
+  {{- end }}
+spec:
+  revisionHistoryLimit: {{ $values.revisionHistoryLimit }}
+  replicas: {{ $values.replicas }}
+  podManagementPolicy: {{ default "OrderedReady" $values.podManagementPolicy }}
+  {{- $strategy := default "RollingUpdate" $values.strategy }}
+  {{- if and (ne $strategy "OnDelete") (ne $strategy "RollingUpdate") }}
+    {{- fail (printf "Not a valid strategy type for StatefulSet (%s)" $strategy) }}
+  {{- end }}
+  updateStrategy:
+    type: {{ $strategy }}
+    {{- if and (eq $strategy "RollingUpdate") $values.rollingUpdate.partition}}
+    rollingUpdate:
+      partition: {{ $values.rollingUpdate.partition }}
+    {{- end }}
+  selector:
+    matchLabels:
+      {{- include "replicated-library.labels.selectorLabels" . | nindent 6 }}
+  serviceName: {{ include "replicated-library.names.fullname" . }}
```

Comment:
This will need to be the name of the headless service that's used with the statefulset. See - https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id. 

So here we'll need to give some thought to if we can help the user automatically set this or only document it. I like the former but I don't think you need to handle that on this PR. I think documenting in the `values.yaml` what `serviceName` is for and defaulting it to the name of the app should be sufficient, and then perhaps create a SC to see if we can have a more user friendly abstraction to automatically attach a headless service to a StatefulSet.

Reply:
What is "SC"

Reply:
Shortcut

Reply:
I am wondering where we should document as it seems like it will get kind of messy and difficult to track if all the documentation is in the values.yaml 

If understand correctly, you want to have a way to specify the serviceName so the end user can define the hostname pattern of the pods to something different that the appName if they want?

Reply:
> I am wondering where we should document as it seems like it will get kind of messy and difficult to track if all the documentation is in the values.yaml

Yeah, there's no better place to document anything right now. So just as everthing else is annotated in the `values.yaml`, you can do the same here and then use the `helm-docs` command from the README, to re-generate everything.

> If understand correctly, you want to have a way to specify the serviceName so the end user can define the hostname pattern of the pods to something different that the appName if they want?

Correct. `serviceName` on the StatefulSet should use `apps.yourappname.serviceName` otherwise default to `yourappname`.